src/io/hdf/hdf_wrapper_functions.h

```diff
@@ -133,6 +133,25 @@ inline bool h5d_read(hid_t grp, const std::string& aname, T* first, hid_t xfer_p
   return ret != -1;
 }
 
+template<typename T>
+inline bool h5d_check_type(hid_t grp, const std::string& aname)
+{
+  if (grp < 0)
+    return true;
+  hid_t h1 = H5Dopen(grp, aname.c_str(), H5P_DEFAULT);
+  if (h1 < 0)
+    return false;
+  T temp;
+  hid_t h5d_type_id     = get_h5_datatype(temp);
+  hid_t datatype        = H5Dget_type(h1);
+  htri_t equality_check = H5Tequal(datatype, h5d_type_id);
```

Comment:
Should we differentiate between `equality_check == 0` (false) and `equality_check < 0` (failure)
https://docs.hdfgroup.org/hdf5/develop/group___h5_t.html#gaa92250f289b557b63cba974defa20b0f

Reply:
Same with the above comment. I think now we can hit an error only if the equality check is done with an `H5I_INVALID_HID` type for the first arg. However I also decided to throw here. The test has been updated with a "REQUIRE_THROWS_AS" test case.